// live/lib/init/__tests__/init-test.js
'use strict';

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var mock = require('mock-fs');
var init = require('..');
var fs = require('fs');

describe('Init', function () {

  beforeEach(function () {

    // For `babel/register`, and webpack.
    var liverc = {
      globals: {},
      moduleDirectories: ['modules']
    };

    var livercStr = 'module.exports = ' + (0, _stringify2.default)(liverc);

    function makeApp() {
      return {
        'package.json': 'TODO',
        '.liverc': livercStr,
        'modules': {
          foo: {
            'live.js': 'module.exports = function(live) { return "foo" }'
          },
          bar: {
            'live.js': 'module.exports = function(live) { return "bar" }'
          }
        }
      };
    }

    // Create test dir structure.
    var testDir = {};
    testDir['/parentApp'] = makeApp();
    testDir['/parentApp/modules/childApp'] = makeApp

    // DEBUG: DISABLED -
    //mock(testDir)

    ();
  });

  it('should work', function () {
    //init()
  });

  afterEach(function () {
    //mock.restore()
  });
});
