// live/lib/init/roadrunner.js
'use strict';

//
// TODO(vjpr): There are two separate caches. Require resolution and babel transpiled source. Separate them!
//

var debug = require('debug')('live');
var roadrunner = require('roadrunner');
var _ = require('lodash');

function getOpts(opts) {
  opts.roadrunner = opts.roadrunner || {};
  _.defaults(opts.roadrunner, {
    roadrunnerCacheFile: process.cwd() + '/tmp/.roadrunner.json'
  });
  return opts;
}

////////////////////////////////////////////////////////////////////////////////

var globalRoadrunnerCacheFilePath = void 0;

var getRequireResolveCache = module.exports.getRequireResolveCache = function () {
  roadrunner.get('babel');
};

var saveRequireResolveCache = module.exports.saveRequireResolveCache = function () {
  roadrunner.save(globalRoadrunnerCacheFilePath);
  debug('Saved roadrunner cache');
};

module.exports = function () {
  var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};


  opts = getOpts(opts);

  globalRoadrunnerCacheFilePath = opts.roadrunnerCacheFilePath;

  var fs = require('fs-extra');

  fs.ensureFileSync(opts.roadrunner.roadrunnerCacheFilePath);

  roadrunner.load(opts.roadrunner.roadrunnerCacheFilePath);

  addSaveOnExitHook();
};

function addSaveOnExitHook() {

  //roadrunner.setup() // We override this function with our own.

  process.on('SIGINT', function () {

    saveRequireResolveCache

    // NOTE: `babel/register` cache hooks into `process.on('exit')`
    // and process.nextTick`.
    // The process exit hook doesnt't fire when we use CMD+C. Not sure why.
    // Maybe something to do with another module, or nodemon, or something.
    // For now, we manually trigger exit which fixes it.

    // TODO: This seems to cause bad exit for nodemon.
    // Maybe try process.emit('exit') ?

    ();process.exit();
  });
}
